ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt

```diff
@@ -243,27 +253,41 @@ fun HttpRequest.parseAuthorizationHeader() = getHeader(Header.AUTHORIZATION)
 /**
  * Throws an exception if the token is invalid. Does nothing otherwise.
  *
- * TODO: is it the good behaviour given:
- * - tokenA: Valid in receiver context, during sender registration (only for sender -> receiver calls)
- * - tokenB: Valid in sender context, during sender registration (only for receiver -> sender calls)
- * - tokenC: Valid when the sender is registered with the receiver (only for sender -> receiver)
- *
  * @throws OcpiClientInvalidParametersException if the token is invalid, otherwise does nothing
  * @throws OcpiClientNotEnoughInformationException if the token is missing
  * @throws HttpException if the authorization header is missing
+ *
  */
-suspend fun PlatformRepository.tokenFilter(httpRequest: HttpRequest) {
+suspend fun PlatformRepository.tokenFilter(
+    httpRequest: HttpRequest
+) {
     val token = httpRequest.parseAuthorizationHeader()
 
-    if (!platformExistsWithTokenA(token) &&
-        !platformExistsWithTokenB(token) &&
-        getPlatformUrlByTokenC(token) == null
-    ) {
-        throw OcpiClientInvalidParametersException("Invalid token: $token")
+    /**
+     * From OCPI 2.2.1 doc:
+     * When a server receives a request with a valid CREDENTIALS_TOKEN_A, on another module than: credentials or
+     * versions, the server SHALL respond with an HTTP 401 - Unauthorized status code.
+     *
+     * So, we allow token A only if we are in this case.
+     */
+
+    val allowTokenA = httpRequest.path.contains("versions") ||
+        httpRequest.path.contains("/{versionNumber}") ||
+        httpRequest.path.contains("credentials")
+
+    val validToken = (allowTokenA && isCredentialsTokenAValid(token)) ||
+        isCredentialsServerTokenValid(token)
+
+    if (!validToken) {
+        throw OcpiClientInvalidParametersException("Invalid server token (token A allowed: $allowTokenA): $token")
     }
 }
 
-suspend fun TransportClientBuilder.buildFor(module: ModuleID, platform: String, platformRepository: PlatformRepository) =
+suspend fun TransportClientBuilder.buildFor(
+    module: ModuleID,
+    platform: String,
```

Comment:
this string is misleading: it's the platformUrl we need here. That's something that is not fully obvious to e yet that we use this platformUrl everywhere to identifiy the platform, I wonder if we could rather use a platform id or `partnerPlatformId`. But this goes beyond the scope of this review

Reply:
fixed